zephyr-build/src/devicetree.rs

```diff
@@ -0,0 +1,291 @@
+    /// The label of our target.  Keep this because it may be useful to know which label was used,
+    /// as nodes often have multiple labels.
+    name: String,
+    /// The inside of the node, inner mutability so this can be looked up and cached.
+    node: RefCell<Option<Rc<Node>>>,
+}
+
+#[derive(Debug)]
+pub enum Word {
+    Number(u32),
+    Phandle(Phandle),
+}
+
+impl DeviceTree {
+    /// Decode the `zephyr.dts` and `devicetree_generated.h` files from the build and build an internal
+    /// representation of the devicetree itself.
+    pub fn new<P1: AsRef<Path>, P2: AsRef<Path>>(dts_path: P1, dt_gen: P2) -> DeviceTree {
+        let ords = OrdMap::new(dt_gen);
+
+        let dts = std::fs::read_to_string(dts_path)
+            .expect("Reading zephyr.dts file");
+        let dt = parse::parse(&dts, &ords);
+
+        // Walk the node tree, fixing any phandles to include their reference.
+        dt.root.phandle_walk(&dt.labels);
+
+        // Walk the node tree, setting each node's parent appropriately.
+        dt.root.parent_walk();
+
+        dt
+    }
+}
+
+impl Node {
+    fn phandle_walk(&self, labels: &BTreeMap<String, Rc<Node>>) {
+        for prop in &self.properties {
+            for value in &prop.value {
+                value.phandle_walk(labels);
+            }
+        }
+        for child in &self.children {
+            child.phandle_walk(labels);
+        }
+    }
+
+    fn parent_walk(self: &Rc<Self>) {
+        for child in &self.children {
+            *(child.parent.borrow_mut()) = Some(self.clone());
+            child.parent_walk()
+        }
+    }
+
+    fn is_compatible(&self, name: &str) -> bool {
+        self.properties
+            .iter()
+            .filter(|p| p.name == "compatible")
+            .flat_map(|prop| prop.value.iter())
+            .any(|v| matches!(v, Value::String(vn) if name == vn))
+    }
+
+    /// A richer compatible test.  Walks a series of names, in reverse.  Any that are "Some(x)" must
+    /// be compatible with "x" at that level.
+    fn compatible_path(&self, path: &[Option<&str>]) -> bool {
+        self.path_walk(path, 0)
+    }
+
+    /// Recursive path walk, to make borrowing simpler.
+    fn path_walk(&self, path: &[Option<&str>], pos: usize) -> bool {
+        if pos >= path.len() {
+            // Once past the end, we consider everything a match.
+            return true;
+        }
+
+        // Check the failure condition, where this node isn't compatible with this section of the path.
+        if matches!(path[pos], Some(name) if !self.is_compatible(name)) {
+            return false;
+        }
+
+        // Walk down the tree.  We have to check for None here, as we can't recurse on the none
+        // case.
+        if let Some(child) = self.parent.borrow().as_ref() {
+            child.path_walk(path, pos + 1)
+        } else {
+            // We've run out of nodes, so this is considered not matching.
+            false
+        }
+    }
+
```

Comment:
```suggestion
    fn path_walk(&self, path: &[Option<&str>]) -> bool {
        match path.first() {
            None => return true;
            Some(current) if matches!(current, Some(name) if !self.is_compatible(name)) => return false;
            _ => ();
        };

        // Walk down the tree.  We have to check for None here, as we can't recurse on the none
        // case.
        if let Some(child) = self.parent.borrow().as_ref() {
            child.path_walk(path[1:])
        } else {
            // We've run out of nodes, so this is considered not matching.
            false
        }
    }

```

Alternately, use slicing to avoid storing and worrying about the position.

Reply:
I've rewritten the `compatible_path` to just call itself with the slice instead of needing a helper.